WL-3899 Dropbox to have username when archived

surya-janani opened the pull request:
1. So far the resource was "eid/filename" which i've altered to be
   "(displayname of the folder+displayId of the user)/filename".
2.  I've concluded that the filename used so far was
      ->"filename" when archived at the student folder level
      ->"studentfolder eid/filename" when archived from root folder level
   Hence I've split file name rather than rootId or resourceId
3. If block handles the case when archived from root folder level and the else block handles the case of archiving from student folder level.

surya-janani committed:
WL-3899 Dropbox to have username

1. So far the resource was "eid/filename" which i've altered to be
"(displayname of the folder+displayId of the user)/filename".
2.  I've concluded that the filename used so far was
       ->"filename" when archived at the student folder level
       ->"studentfolder eid/filename" when archived from root folder level
Hence I've split file name rather than rootId or resourceId

3. If block handles the case when archived from root folder level and the else block handles the case of archiving from student folder level.

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +401,36 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+				ResourcePropertiesEdit props = collectionEdit.getPropertiesEdit();
+				String displayName = props.getProperty(ResourcePropertiesEdit.PROP_DISPLAY_NAME);
+				try {
+					String uniqueId = UserDirectoryService.getUser(filenameArr[0]).getDisplayId();
```
Does this end up doing an user lookup for every file included in the export? Although there is heavy caching of found users this will query the DB for every file I think?

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +401,36 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+		if(filename != null && filename.length()>0) {
+			String filenameArr[] = filename.split(Entity.SEPARATOR);
+			if (filenameArr.length > 1) {
+				ContentCollectionEdit collectionEdit = (ContentCollectionEdit) ContentHostingService.getCollection(rootId+filenameArr[0]+Entity.SEPARATOR);
```
You shouldn't have all this duplicated code. (https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +401,36 @@ private void storeContentCollection(String rootId, ContentCollection collection,
 	 */
-	private void storeContentResource(String rootId, ContentResource resource, ZipOutputStream out) throws Exception {		
-		String filename = resource.getId().substring(rootId.length(),resource.getId().length());				
+	private void storeContentResource(String rootId, ContentResource resource, ZipOutputStream out) throws Exception {
```
Isn't this the same code that is used for compressing folders from the resources tool as well? If you have a folder/file that has the same ID as a user then it will get a strange name in the export.

surya-janani committed:
WL-3899 Changing names only when the request is from group-user(dropbox)
Making zipped folder have user names rather than eid's

surya-janani committed:
WL-3899 Refactoring the code

surya-janani commented:
Adding ServerConfiguration validation and refactoring

surya-janani commented:
Linked with  https://github.com/ox-it/wl-content/commit/41882f3531fbecedc9424c33b43d458d4d2556c8

surya-janani committed:
WL-3899 Making displayname changes in the methods where dropbox folder is created

surya-janani committed:
WL-3899 Making null check before setting displayId

surya-janani commented:
Adding null check before setting displayId

surya-janani committed:
WL-3899 .Removing user directory look up ,
It is redundant with the look up in base content service.

surya-janani commented:
Commenting again-Removing user directory look up in Zip content util as it is redundant with the changes in Basecontentservice

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -36,6 +37,10 @@
 import org.sakaiproject.exception.TypeException;
 import org.sakaiproject.tool.api.ToolSession;
 import org.sakaiproject.tool.cover.SessionManager;
+import org.sakaiproject.user.api.User;
```
Unused imports.

buckett commented on kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java:
```diff
@@ -9588,7 +9588,14 @@ public String getDropboxDisplayName(String siteId)
 
 		// return the current user's sort name
-		return userDirectoryService.getCurrentUser().getSortName();
+		String displayId = userDirectoryService.getCurrentUser().getDisplayId();
```
Same chunk of code in 3 places, it should be refactored out into a method taking a user parameter. That makes it easier if someone needs to change it in the future.

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -121,6 +128,18 @@ public void compressFolder(Reference reference) {
 					newResourceName += ZIP_EXTENSION;
+
+					ContentCollectionEdit currentEdit;
+					if(reference.getId().split(Entity.SEPARATOR).length>3) {
```
So this changes it so that the displayname for the created zipfile is the displayname of the folder with a .zip extension?
And this change affects all zipfiles created, rather than just ones in the dropbox?

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +413,34 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+
+		//Inorder to have username as the folder name rather than having eids
+		if(rootId.indexOf("group-user")!=-1 && ServerConfigurationService.getBoolean("dropbox.zip.haveDisplayname", true)) {
+			if (filename != null && filename.length() > 0) {
```
Why would the filename ever be null or zero length? If it is then it will already throw a NullPointerException when creating the ZipEntry.

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +413,34 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+						filename = displayName + Entity.SEPARATOR + filenameArr[0];
+					}
+				} catch (Exception e) {
+					LOG.warn("Unexpected error occurred when trying to create Zip archive:" + extractName(rootId));
```
If you're logging a warn and it's not expected in normal operations, it's useful to include the stack trace as it makes debugging the problem much easier.

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +413,34 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+						displayName = props.getProperty(ResourcePropertiesEdit.PROP_DISPLAY_NAME);
+						filename = displayName + Entity.SEPARATOR + filenameArr[0];
+					}
+				} catch (Exception e) {
```
You shouldn't generally catch Exception as it's too general. If you want to just deal with the Exceptions from ContentHostingService then you can catch SakaiException, but if you can't read one of the files then you probably want to stop building the zipfile anyway as silently swallowing the error and having an zipfile that doesn't contain what it should is worse in my opinion.

surya-janani committed:
WL-3899 Refactoring code and improving exception handling

surya-janani commented:
I first refactored the code in BaseContentService , and thought I could refactor code in ZipContentUtil in a simmilar way. So I've refactored both and changed exception handling as well .Let me know what you think.

buckett commented on kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java:
```diff
@@ -13868,5 +13869,16 @@ private String getMacroValue(String macroName) {
 		return macroName;
 	}
 
+	private String getDisplayName(User userIn,boolean isUserdir) {
```
`isUserdir` seems redundant as can't you always pass in the user you are want to build the display name for?

surya-janani committed:
WL-3899 Removing isUserdir parameter

surya-janani commented:
Removing isUserdir param

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +408,26 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+		String filename = resource.getId().substring(rootId.length(),resource.getId().length());
+		//Inorder to have username as the folder name rather than having eids
+		if(rootId.indexOf("group-user")!=-1 && ServerConfigurationService.getBoolean("dropbox.zip.haveDisplayname", true)) {
+				try {
```
Indentation looks wrong here.

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +408,26 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+	private void storeContentResource(String rootId, ContentResource resource, ZipOutputStream out) throws Exception {
+		String filename = resource.getId().substring(rootId.length(),resource.getId().length());
+		//Inorder to have username as the folder name rather than having eids
+		if(rootId.indexOf("group-user")!=-1 && ServerConfigurationService.getBoolean("dropbox.zip.haveDisplayname", true)) {
```
There's a method todo this test of `ContentHostingService.isInDropbox(String)`.

buckett commented on kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java:
```diff
@@ -394,8 +408,26 @@ private void storeContentCollection(String rootId, ContentCollection collection,
+				try {
+					filename = getContainingFolderDisplayName(rootId, filename);
+				} catch(TypeException e){
+					LOG.warn("Unexpected error occurred when trying to create Zip archive:" + extractName(rootId)+e.getStackTrace());
```
If it's something the developer needs to debug, you're much better off passing the stack trace as a second argument to the log call, ie `LOG.warn(String,Throwable)`.

surya-janani committed:
WL-3899 Changing exception handling to string,throwable

surya-janani commented:
Changing exception handling to string,throwable

buckett commented:
There are still other comments not resolved on this.

surya-janani committed:
WL-3899 Using ContentHostingServic.isInDropBox and indendation

surya-janani commented:
Using ContentHostingServic.isInDropBox and indendation
